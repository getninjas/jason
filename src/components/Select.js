import PropTypes from 'prop-types';
import React, { Component } from 'react';

const propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onFieldChange: PropTypes.func.isRequired,
  onFieldBlur: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  required: PropTypes.bool.isRequired,
  selected: PropTypes.any,
  initialValue: PropTypes.any,
  values: PropTypes.array,
  type: PropTypes.string,
  placeholder: PropTypes.string,
  style: PropTypes.string,
};

const defaultProps = {
  id: '',
  name: '',
  title: '',
  required: false,
  initialValue: '',
  values: [],
  placeholder: '',
  style: 'form__input',
};

export default class Select extends Component {
  constructor(props) {
    super(props);

    this.state = {
      values: [],
      value: this.props.initialValue ? this.props.initialValue : '',
    };

    this.onChange = this.onChange.bind(this);
    this.onBlur = this.onBlur.bind(this);
  }

  onChange(evt) {
    this.setState({ value: evt.target.value });

    this.props.onFieldChange({
      value: evt.target.value,
      id: this.props.id,
    });
  }

  onBlur() {
    this.props.onFieldBlur({
      value: this.state.value,
      id: this.props.id,
      required: this.props.required,
      type: this.props.type,
    });
  }

  render() {
    const { id, name, required, style } = this.props;

    return (
      <select
        id={id}
        name={name}
        value={this.state.value}
        className={style}
        onChange={this.onChange}
        onBlur={this.onBlur}
        required={required ? 'true' : 'false'}>
        {
          this.props.values.map((item, index) =>
            (
              <option key={`${id}-${index}`} value={item.databaseId}>
                {item.value}
              </option>
            ),
          )
        }
      </select>
    );
  }
}

Select.propTypes = propTypes;
Select.defaultProps = defaultProps;
