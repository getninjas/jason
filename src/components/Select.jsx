import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class Select extends Component {
  constructor(props) {
    super(props);

    this.state = {
      values: []
    }
  }

  componentDidMount() {
    const values = this.addPlaceholder(this.props);

    this.setState({values});
  }

  addPlaceholder({ values, placeholder }) {
    const localValues = [...values];

    placeholder.length? localValues.unshift({databaseId: '', value: placeholder}) : '';

    return localValues;
  }

  render() {
    const { id, name, selected, required } = this.props;

    return (
      <select
        id={id}
        name={name}
        defaultValue={selected}
        className="form__input"
        required={required ? 'true' : 'false'}>
        {
          this.state.values.map((item, index) => {
            return (
              <option key={`option-${index}`} value={item.databaseId}>
                {item.value}
              </option>
            );
          })
        }
      </select>
    );
  }
}

Select.defaultProps = {
  id: '',
  selected: '',
  name: '',
  title: '',
  required: false,
  values: [],
  placeholder: '',
}

Select.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  required: PropTypes.bool.isRequired,
  selected: PropTypes.any,
  values: PropTypes.array,
  placeholder: PropTypes.string,
}
