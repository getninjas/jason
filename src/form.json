{
  "form": {
    "title": "Form Title",
    "databaseId": 81,
    "zipcodeUrlService": "http://www.mocky.io/v2/5afaebc92f00000e00f7c654",
    "steps": [
      {
        "title": "General data",
        "breadcrumbTitle": "Your request",
        "buttonText": "Start request",
        "headerMarkup": "<h2 class='widget__title section-title--high-contrast'>Lorem Ipsum is simply dummy text <small>It has survived not</small></h2>",
        "fields": [
          {
            "title": "Qual é o tipo de serviço que você precisa?",
            "placeholder": "Placeholder",
            "required": true,
            "id": "1_id",
            "name": "1_name_select",
            "type": "select",
            "value": 7117,
            "values": [
              {
                "databaseId": 7117,
                "value": "Construção"
              },
              {
                "databaseId": 7118,
                "value": "Reformas"
              },
              {
                "databaseId": 7119,
                "value": "Instalações"
              },
              {
                "databaseId": 7120,
                "value": "Troca"
              },
              {
                "databaseId": 7121,
                "value": "Outros"
              }
            ]
          },
          {
            "title": "Especifique melhor a metragem do serviço",
            "required": true,
            "id": "2_id",
            "name": "2_name_text",
            "type": "text",
            "placeholder": "Bora pra action",
            "value": "AHUAHuhhHUHAU",
            "values": []
          },
          {
            "title": "Informações adicionais",
            "required": false,
            "type": "textarea",
            "id": "3_id",
            "name": "3_name_textarea",
            "placeholder": "Descreva o que você precisa",
            "value": null,
            "values": []
          },
          {
            "required": true,
            "title": "Você precisa de serviços adicionais?",
            "placeholder": "Você precisa de serviços adicionais?",
            "id": "0_id",
            "name": "1_name_select",
            "type": "checkbox",
            "value": null,
            "values": [
              {
                "databaseId": 6792,
                "value": "Cozinhar"
              },
              {
                "databaseId": 7639,
                "value": "OTHER"
              }
            ]
          }
        ]
      },
      {
        "title": "User Data",
        "breadcrumbTitle": "Your data",
        "buttonText": "Finish",
        "headerMarkup": "<div class='form__alert-message space-element-large'><img src='https://www.getninjas.com.br/assets/request/phone_call.svg' class='form__icon' alt='Lorem Ipsum is simply dummy text of the printing and typesetting industry.'><h2 class='form__info'>Almost there<small>Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's<strong>standard dummy text ever since the 1500s</strong></small></h2></div>",
        "fields": [
          {
            "title": "Zipcode",
            "required": true,
            "type": "zipcode",
            "id": "zip_code",
            "name": "zipcode",
            "placeholder": "00000-000",
            "value": "04904-060",
            "wrapperClassName": "form__field input",
            "values": []
          },
          {
            "title": "DDD + Celular",
            "id": "phone",
            "name": "phone",
            "type": "phone",
            "placeholder": "(__) _____-____",
            "required": true,
            "value": "(11) 96311-2689",
            "values": []
          }
        ]
      }
    ]
  }
}
