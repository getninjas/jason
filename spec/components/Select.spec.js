import React from 'react';
import renderer from 'react-test-renderer';
import Select from '../../src/components/Select';
import { form } from '../../src/form.json';
import { enzymeConfig, shallow, mount } from '../enzymeConfig';

enzymeConfig();

describe('Select', () => {
  it('renders defaultProps', () => {
    const values = form.steps[0].fields[0].values.slice(0, 2);

    const component = renderer.create(
      <Select
        id={'idTest'}
        name={'nameTest'}
        placeholder={'placeholderTest'}
        required={false}
        onFieldChange={() => {}}
        values={values}
      />,
    );

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  describe('.onChange', () => {
    it('updates state selected', () => {
      const values = form.steps[0].fields[0].values.slice(0, 2);

      const component = shallow(
        <Select
          id={'idTest'}
          name={'nameTest'}
          placeholder={'placeholderTest'}
          required={false}
          onFieldChange={() => {}}
          values={values}
        />,
      );

      component.simulate('change', { target: { value: 'Reformas' } });

      expect(component.instance().state.value).toEqual('Reformas');
    });
  });

  describe('.addPlaceholder', () => {
    const values = form.steps[0].fields[0].values.slice(0, 2);

    it('sets placeholder as first option', () => {
      const component = mount(
        <Select
          id={'idTest'}
          name={'nameTest'}
          placeholder={'placeholderTest'}
          required={false}
          onFieldChange={() => {}}
          values={values}
        />,
      );

      expect(component.instance().state.values[0].value).toEqual('placeholderTest');
    });

    it('does not set placeholder', () => {
      const component = mount(
        <Select
          id={'idTest'}
          name={'nameTest'}
          placeholder={''}
          required={false}
          onFieldChange={() => {}}
          values={values}
        />,
      );

      expect(component.instance().state.values[0].value).not.toEqual('');
    });
  });
});
