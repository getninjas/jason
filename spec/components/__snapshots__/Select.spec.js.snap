// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders defaultProps 1`] = `
<select
  className="form__input"
  defaultValue=""
  id="idTest"
  name="nameTest"
  onBlur={[Function]}
  onChange={[Function]}
  required="false"
>
  <option
    value=""
  >
    placeholderTest
  </option>
  <option
    value={7117}
  >
    Construção
  </option>
  <option
    value={7118}
  >
    Reformas
  </option>
</select>
`;
